pkg/inspector/image-inspector.go

```diff
@@ -467,3 +414,153 @@ func createOutputDir(dirName string, tempName string) (string, error) {
+	if err != nil {
+		return nil, "", err
+	}
+	defer policyContext.Destroy()
+
+	srcRef, err := alltransports.ParseImageName(i.opts.Image)
+	if err != nil {
+		// try adding "docker://"
+		i.opts.Image = "docker://" + i.opts.Image
+		srcRef, err = alltransports.ParseImageName(i.opts.Image)
+		if err != nil {
+			return nil, "", fmt.Errorf("Invalid source name %s: %v", i.opts.Image, err)
+		}
+	}
+
+	certPath, err := i.certPath(i.opts.Image)
+	if err != nil {
+		return nil, "", err
+	}
+	sourceCtx := &types.SystemContext{
+		DockerAuthConfig: nil,
+		DockerCertPath:   certPath,
+	}
+
+	if i.opts.DstPath, err = createOutputDir(i.opts.DstPath, "image-inspector-"); err != nil {
+		return nil, "", err
+	}
+	destRef, err := alltransports.ParseImageName(fmt.Sprintf("dir://%s", i.opts.DstPath))
+	if err != nil {
+		return nil, "", fmt.Errorf("Invalid destination name %s: %v", i.opts.DstPath, err)
+	}
+
+	imagePullAuths, err := i.getAuthConfigs()
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Try all the possible auth's from the config file
+	var authErr error
+	log.Println("Copying image")
+	for name, auth := range imagePullAuths.Configs {
+		sourceCtx.DockerAuthConfig = &types.DockerAuthConfig{
+			Username: auth.Username,
+			Password: auth.Password,
+		}
+		authErr = copy.Image(policyContext, destRef, srcRef, &copy.Options{
+			RemoveSignatures: false,
+			SignBy:           "",
+			ReportWriter:     os.Stdout,
+			SourceCtx:        sourceCtx,
+			DestinationCtx:   nil,
+			ProgressInterval: PULL_LOG_INTERVAL_SEC,
+		})
+		if authErr == nil {
+			break
+		}
+		log.Printf("Authentication with %s failed: %v", name, authErr)
+	}
+
+	if authErr != nil {
+		return nil, "", fmt.Errorf("Unable to pull docker image: %v\n", authErr)
+	}
+
+	img, err := srcRef.NewImage(sourceCtx)
+	if err != nil {
+		return nil, "", err
+	}
+
+	log.Println("Inspecting image")
+	rawManifest, _, err := img.Manifest()
+	if err != nil {
+		return nil, "", fmt.Errorf("Error while reading image manifest: %v\n", err)
+	}
+	imageDigest, err := manifest.Digest(rawManifest)
+	if err != nil {
+		return nil, "", fmt.Errorf("Error while parsing image manifest: %v\n", err)
+	}
+
+	inspectInfo, err := img.Inspect()
+	if err != nil {
+		return inspectInfo, "", fmt.Errorf("Error while inspecting copied image Manifest: %v\n", err)
+	}
+
+	err = i.extractDownloadedImage(inspectInfo)
+	if err != nil {
+		return nil, "", fmt.Errorf("Error while extracting downloaded image: %v\n", err)
+	}
+
+	return inspectInfo, imageDigest, nil
+}
+
+// extractDownloadedImage will untar all the layer tar files specified in 'info'
+// assuming those files exist in i.opts.DstPath .
+func (i *defaultImageInspector) extractDownloadedImage(info *types.ImageInspectInfo) error {
+	for _, layer := range info.Layers {
+		filename := i.opts.DstPath + "/" + strings.SplitN(layer, ":", 2)[1] + ".tar"
+		log.Printf("Untar %s\n", filename)
+		if err := util.UntarGzFile(filename, i.opts.DstPath); err != nil {
```

Comment:
Is this sufficient? The tarfile may contain whiteouts ( https://github.com/opencontainers/image-spec/blob/master/layer.md#whiteouts ), and I can’t see anything handling that.

Or perhaps this is irrelevant for your use case?